adapters/showheroes/showheroes.go

```diff
@@ -0,0 +1,228 @@
+		}
+	}
+
+	if request.Site == nil && request.App == nil {
+		return &errortypes.BadInput{
+			Message: "request must contain a site or an app",
+		}
+	}
+
+	return nil
+}
+
+func processImp(imp *openrtb2.Imp, reqInfo *adapters.ExtraRequestInfo) error {
+	var bidderExt shExtImpBidder
+	if err := jsonutil.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: "Error parsing bidder params",
+		}
+	}
+
+	// move params from .bidder to .params
+	// this is required since openrtb_ext.ExtImpShowheroes is used in 2 places:
+	// 1. shExtImpBidder.Bidder - for parsing the incoming request
+	// 2. shExtImpBidder.Params - for marshaling the outgoing request to showheroes
+	bidderExt.Params.UnitID = bidderExt.Bidder.UnitID
+
+	impExt, err := jsonutil.Marshal(bidderExt)
+	if err != nil {
+		return err
+	}
+	imp.Ext = impExt
+
+	if imp.BidFloor == 0 || imp.BidFloorCur == "EUR" {
+		return nil
+	}
+
+	// convert the bid floor to EUR
+	currency := imp.BidFloorCur
+	// default currency according to the openRTB is USD
+	if currency == "" {
+		currency = "USD"
+	}
+
+	eurFloor, err := reqInfo.ConvertCurrency(imp.BidFloor, currency, "EUR")
+	if err != nil {
+		return err
+	}
+
+	imp.BidFloor = eurFloor
+	imp.BidFloorCur = "EUR"
+
+	return nil
+}
+
+func getPrebidChannel(request *openrtb2.BidRequest) (string, string) {
+	var channelName string
+	var channelVersion string
+	reqExt := &openrtb_ext.ExtRequest{}
+
+	if err := jsonutil.Unmarshal(request.Ext, reqExt); err == nil && reqExt.Prebid.Channel != nil {
+		channelName = reqExt.Prebid.Channel.Name
+		channelVersion = reqExt.Prebid.Channel.Version
+	}
+	return channelName, channelVersion
+}
+
+func setPBSVersion(request *openrtb2.BidRequest, pbsVersion string) {
+	if pbsVersion == "" {
+		return
+	}
+	source := request.Source
+	if source == nil {
+		source = &openrtb2.Source{}
+		request.Source = source
+	}
+
+	sourceExtMap := make(map[string]json.RawMessage)
+	if source.Ext != nil {
+		if err := jsonutil.Unmarshal(source.Ext, &sourceExtMap); err != nil {
+			// if we can't parse the existing ext, don't modify it
+			return
+		}
+	}
+	sourceExtMap["pbs"] = json.RawMessage(`{"pbsv":"` + pbsVersion + `", "pbsp": "go"}`)
+	if ext, err := json.Marshal(sourceExtMap); err == nil {
+		source.Ext = ext
+	}
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, extra *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if err := validate(request); err != nil {
+		return nil, []error{err}
+	}
+	var errors []error
+	validImps := make([]openrtb2.Imp, 0, len(request.Imp))
+
+	prebidChannelName, channelVersion := getPrebidChannel(request)
+
+	// pre-process the imps
+	for _, imp := range request.Imp {
+		if err := processImp(&imp, extra); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		// if display manager is not set and request came from prebid.js
+		// store it and its version
+		if imp.DisplayManager == "" {
```

Comment:
From the comment in lines 152 and 153 it seems that `prebidChannelName` must equal `"pbjs"` to get copied to `imp.DisplayManager`, am I wrong? Even if that's not the case, should we validate `prebidChannelName` and `channelVersion` for non-emptiness before assigning?
```
150         }
151
152         // if display manager is not set and request came from prebid.js
153         // store it and its version
154 -       if imp.DisplayManager == "" {
    +       if imp.DisplayManager == "" && prebidChannelName == "pbjs" {
155             imp.DisplayManager = prebidChannelName
156             imp.DisplayManagerVer = channelVersion
157         }
158         validImps = append(validImps, imp)
159     }
adapters/showheroes/showheroes.go
```
If not, I'm ok either way.